libsplinter/src/auth/oauth/builder/mod.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Builders for [OAuthClient](crate::auth::oauth::OAuthClient) structs.
+
+#[cfg(feature = "oauth-github")]
+mod github;
+#[cfg(feature = "oauth-openid")]
+mod openid;
+
+use crate::auth::rest_api::identity::IdentityProvider;
+use crate::error::InvalidStateError;
+
+use super::error::OAuthClientBuildError;
+use super::OAuthClient;
+
+#[cfg(feature = "oauth-github")]
+pub use github::GithubOAuthClientBuilder;
+
+#[cfg(feature = "oauth-openid")]
+pub use openid::OpenIdOAuthClientBuilder;
+
+/// A builder for a new [`OAuthClient`].
+///
+/// This builder constructs an [`OAuthClient`] using the most general parameters. Configurations
+/// that set values specific to certain providers may be available, depending on which features
+/// have been enabled at compile time.
+#[derive(Default)]
+pub struct OAuthClientBuilder {
+    client_id: Option<String>,
+    client_secret: Option<String>,
+    auth_url: Option<String>,
+    redirect_url: Option<String>,
+    token_url: Option<String>,
+    scopes: Vec<String>,
+    identity_provider: Option<Box<dyn IdentityProvider>>,
+}
+
+impl OAuthClientBuilder {
+    /// Constructs a new builder.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Constructs a new builder preconfigured with settings for the Github OAuth2 provider.
```

Comment:
These new_ methods seem out-of-place since they aren't constructors in this context. Their use as constructor-looking functions results in misleading code (you would expect a new_* to return OAuthClientBuilder, but even though that's how the code reads, that's not what is going on).

Reply:
Replaced with the individual constructors